### PR TITLE
Fix exception thrown when inserting new line with toggled inline styles

### LINKED ISSUE #159: Inserting new line when inline styles like bold is toggled causes exception
## Steps to Reproduce
1. Toggle inline style like bold
2. Press enter

## Environment
- Flutter version 3.13.2
- Fleather version 1.10.0

## Logs

```console
══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
The following assertion was thrown during method call TextInputClient.updateEditingStateWithDeltas:
Assertion failed:
file:///fleather/packages/parchment/lib/src/document/line.dart:245:11
line.dart:245
style.values
              .every((attr) => attr.scope == ParchmentAttributeScope.line)
"It is not allowed to apply inline attributes to line itself."
When the exception was thrown, this was the stack:
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:49      throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 35:3        assertFailed
packages/parchment/src/document/line.dart 246:70                                  retain
packages/parchment/src/document/node.dart 252:8                                   retain
packages/parchment/src/document.dart 250:15                                       compose
packages/fleather/src/widgets/controller.dart 106:18                              replaceText
packages/fleather/src/widgets/editor_input_client_mixin.dart 136:25               updateEditingValueWithDeltas
packages/flutter/src/services/text_input.dart 1864:10                             _handleTextInputInvocation
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/flutter/src/services/text_input.dart 1768:45                             [_handleTextInputInvocation]
packages/flutter/src/services/text_input.dart 1753:20                             _loudlyHandleTextInputInvocation
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/flutter/src/services/text_input.dart 1751:51                             [_loudlyHandleTextInputInvocation]
packages/flutter/src/services/platform_channel.dart 547:55                        _handleAsMethodCall
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/flutter/src/services/platform_channel.dart 544:40                        [_handleAsMethodCall]
packages/flutter/src/services/platform_channel.dart 540:34                        <fn>
packages/flutter/src/services/binding.dart 567:35                                 <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/flutter/src/services/binding.dart 564:98                                 <fn>
lib/_engine/engine/platform_dispatcher.dart 1335:13                               invoke2
lib/ui/channel_buffers.dart 25:12                                                 invoke
lib/ui/channel_buffers.dart 65:7                                                  push
lib/ui/channel_buffers.dart 131:16                                                push
lib/_engine/engine/platform_dispatcher.dart 438:25                                invokeOnPlatformMessage
lib/_engine/engine/text_editing/text_editing.dart 2182:39                         updateEditingStateWithDelta
lib/_engine/engine/text_editing/text_editing.dart 2289:19                         <fn>
lib/_engine/engine/text_editing/text_editing.dart 1378:7                          handleChange
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall
call:
  MethodCall(TextInputClient.updateEditingStateWithDeltas, [1, {deltas: [{oldText: Test
  , deltaText:
  , deltaStart: 4, deltaEnd: 4, selectionBase: 5, selectionExtent: 5, composingBase: -1,
  composingExtent: -1}]}])
════════════════════════════════════════════════════════════════════════════════════════════════════
```

### PR DESCRIPTION
**Pull request by Amir-P**
Fixes #159

### CODE REVIEW

File: packages/fleather/lib/src/widgets/controller.dart [hunk 1]
```diff
@@ -69,14 +70,40 @@ class FleatherController extends ChangeNotifier {
...
-          ));
+  bool _shouldApplyToggledStyles(Delta delta) {
+    if (toggledStyles.isNotEmpty && delta.isNotEmpty) {
+      // covers single insert and a retain+insert
+      if (delta.length <= 2 && delta.last.isInsert) {
+        return true;
+      }
+      // special case for AutoTextDirectionRule
+      if (delta.length <= 3 && delta.last.isRetain) {
+        return delta.last.attributes
```
Comment (amantoux) [inline packages/fleather/lib/src/widgets/controller.dart:81]: Wouldn't it be easier to read like this?
```dart
return delta.last.attributes != null && delta.last.attributes!.contains... && ....;
```

WDYT?

File: packages/fleather/lib/src/widgets/controller.dart [hunk 2]
```diff
@@ -69,14 +70,40 @@ class FleatherController extends ChangeNotifier {
...
-              delta.last.isRetain // special case for AutoTextDirectionRule
-          ));
+  bool _shouldApplyToggledStyles(Delta delta) {
+    if (toggledStyles.isNotEmpty && delta.isNotEmpty) {
+      // covers single insert and a retain+insert
+      if (delta.length <= 2 && delta.last.isInsert) {
+        return true;
+      }
+      // special case for AutoTextDirectionRule
+      if (delta.length <= 3 && delta.last.isRetain) {
```
Comment (amantoux) [inline packages/fleather/lib/src/widgets/controller.dart:80]: I fail to understand which special case it is meant to cover
Any idea?

File: packages/fleather/lib/util.dart [hunk 3]
```diff
@@ -53,3 +53,8 @@ TextDirection getDirectionOfNode(StyledNode node) {
   }
   return TextDirection.ltr;
 }
+
+bool isDataOnlyNewLines(Object data) {
+  if (data is! String) return false;
```
Comment (amantoux) [inline packages/fleather/lib/util.dart:58]: Add quick corner case
```dart
if (data.isEmpty) return false;
``````

File: packages/fleather/lib/util.dart [hunk 4]
```diff
@@ -53,3 +53,8 @@ TextDirection getDirectionOfNode(StyledNode node) {
...
+  return data.replaceAll('\n', '').isEmpty;
```
Comment (amantoux) [inline packages/fleather/lib/util.dart:59]: Don't you think this could be checked quicker and without any memory overhead with a regex?

**Review by amantoux — CHANGES_REQUESTED** [review]
LGTM overall, some nitpicks

File: packages/fleather/lib/src/widgets/controller.dart (hunk 1 above)
Comment (Amir-P) [reply packages/fleather/lib/src/widgets/controller.dart:81]: Agreed.

File: packages/fleather/lib/src/widgets/controller.dart (hunk 2 above)
Comment (Amir-P) [reply packages/fleather/lib/src/widgets/controller.dart:80]: Check this out: https://github.com/fleather-editor/fleather/blob/187ac7af156d22687a0fe60913222bd070ef76d6/packages/parchment/lib/src/heuristics/insert_rules.dart#L620

**Review by amantoux — APPROVED** [review]
All good for here

**Pull request merged**